daemon/SecretsImpl/applicationpermissions.cpp

```diff
@@ -12,46 +12,125 @@
 #include <QtCore/QDir>
 #include <QtCore/QFileInfo>
 
-QString Sailfish::Secrets::Daemon::ApiImpl::ApplicationPermissions::applicationId(pid_t pid) const
-{
-    // TODO: readlink /proc/pid/exe -- but requires root permissions?
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+namespace {
+    QString readBoosterCgroup(pid_t pid)
+    {
+        const QString pidFile(QStringLiteral("/proc/%1/cgroup").arg(pid));
+        if (!QFile::exists(pidFile)) {
+            qCDebug(lcSailfishSecretsDaemon) << "no cgroup in procfs for process:" << pid;
+            return QString();
+        }
+
+        QFile file(pidFile);
+        if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+            qCDebug(lcSailfishSecretsDaemon) << "unable to open cgroup file:" << pidFile;
+            return QString();
+        }
+
+        const QList<QByteArray> lines(file.readAll().split('\n'));
+        for (const QByteArray &line : lines) {
+            int idx = line.indexOf(":name=booster:");
+            if (idx >= 0) {
+                const QByteArray cgroupName(line.mid(idx + strlen(":name=booster:")));
+                if (!cgroupName.isEmpty() && cgroupName != QStringLiteral("/")) {
+                    const QString retn(QString::fromUtf8(cgroupName).trimmed());
+                    qCDebug(lcSailfishSecretsDaemon) << "caller with pid" << pid << "has cgroup applicationId:" << retn;
+                    return retn;
+                }
+            }
+        }
 
-    // special case handling of the pid 999999 for demonstration testing purposes only!
-    // TODO: remove this special case handling!
-    if (pid == 999999) {
-        qWarning() << "WARNING! Special case pid handling needs to be removed once access control / application management API exists!";
-        return QLatin1String("test-third-party-application");
+        return QString();
     }
 
+    QString readExeSymlink(pid_t pid)
+    {
+        struct stat sb;
+        const QString pidFile(QStringLiteral("/proc/%1/exe").arg(pid));
+        if (lstat(pidFile.toStdString().c_str(), &sb) == -1) {
```

Comment:
Maybe use `pidFile.toUtf8().data()` that way it's clearer what is going on.

Reply:
done